### DISCUSSION TITLE
Adding Crashlytics Logger as An appender

### DISCUSSION THREAD

**Discussion by ismdcf** [tony19/logback-android#370] [Ideas]
Is it possible to add Crashlytics Logger as an appender. So that when there is a crash on Crashlytics we have the latest logs which were logged using the logback.

**Comment by azabost** [comment]
Did you try something like this?

```kotlin
class CrashlyticsAppender(
    private val encoder: PatternLayoutEncoder
) : AppenderBase<ILoggingEvent>() {

    override fun append(eventObject: ILoggingEvent) {
        try {
            FirebaseCrashlytics.getInstance().log(encoder.layout.doLayout(eventObject))
        } catch (e: Exception) {
            addError("Failed to log to Crashlytics", e)
        }
    }
}
```

It may require some adjustments, e.g. enabling/disabling Crashlytics in runtime based on user's choice or making sure that `Firebase.initializeApp()` is called first.

And then you must add the appender. I haven't used the XML configs so I can only explain how it can be done programmatically instead:

```kotlin
val lc = LoggerFactory.getILoggerFactory() as ch.qos.logback.classic.LoggerContext

val crashlyticsEncoder = PatternLayoutEncoder()
crashlyticsEncoder.context = lc
crashlyticsEncoder.pattern = "[%thread] %msg%n" // modify this as you wish
crashlyticsEncoder.start()

val crashlyticsAppender = CrashlyticsAppender(crashlyticsEncoder)
crashlyticsAppender.context = lc
crashlyticsAppender.start()

val root = LoggerFactory.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME) as ch.qos.logback.classic.Logger
root.addAppender(crashlyticsAppender)
```

Note it's only a snippet and not a full configuration. Please follow some examples that you can easily find elsewhere e.g.

https://stackoverflow.com/questions/16910955/programmatically-configure-logback-appender
https://akhikhl.wordpress.com/2013/07/11/programmatic-configuration-of-slf4jlogback/
